### ISSUE TITLE
Malformed Nalus near timestamp 51900, NaluSize > DataSize!

### ISSUE THREAD

**Issue by jkuri** [Bilibili/flv.js#273]
Hi! First of all, thanks for this awesome library!

I am streaming flv with a ffmpeg c++ api and my code looks like [that](https://github.com/jkuri/opencv-ffmpeg-rtmp-stream/blob/master/src/rtmp-stream.cpp).
When trying to play video with `flv.js` I am getting errors
```
Flv: Invalid AVCDecoderConfigurationRecord, lack of data!
Malformed Nalus near timestamp 51900, NaluSize > DataSize!
Malformed Nalus near timestamp 51901, NaluSize > DataSize!
Malformed Nalus near timestamp 51902, NaluSize > DataSize!
...
```
but I can play video with `ffplay` normally.
where is the catch? would appreciate any help. thanks!

**Comment by LanderlYoung** [comment]
The same problem occurred to me.
not currently solved, but I think it the flv file format issue.

**Comment by jkuri** [comment]
@LanderlYoung I managed to get video appear in the browser.
After two days of searching for a solution this is what makes things work:
```cpp
unsigned char sps_pps[23] = { 0x00, 0x00, 0x00, 0x01, 0x67, 0x42, 0x00, 0x0a, 0xf8, 0x41, 0xa2, 0x00, 0x00, 0x00, 0x01, 0x68, 0xce, 0x38, 0x80 };
out_codec_ctx->extradata_size = 23;
out_codec_ctx->extradata = (uint8_t*)av_malloc(23 + AV_INPUT_BUFFER_PADDING_SIZE);
memcpy(out_codec_ctx->extradata, sps_pps, 23);
```
but there's a new catch, video works really slow.

**Comment by jkuri** [comment]
please let me know if you find a solutoion for that.

**Comment by xqq** [comment]
Why hardcode `sps and pps`?
This structure must be generated dynamically from your exact stream parameters.

https://github.com/jkuri/opencv-ffmpeg-rtmp-stream/blob/master/src/rtmp-stream.cpp#L213
Simply increase `PTS` by 1 is totally wrong. Obviously **Presentation Time Stamp** is related to the fps of video stream.

**Comment by jkuri** [comment]
@xqq I appreciate you help! 
I noticed if I set

```cpp
codec_ctx->flags |= AV_CODEC_FLAG_GLOBAL_HEADER;
```

and copy `extradata` from dynamically generated sps and pps (might be a bug in ffmpeg that this is not done automatically)

```cpp
out_stream->codecpar->extradata = out_codec_ctx->extradata;
out_stream->codecpar->extradata_size = out_codec_ctx->extradata_size;
```

I can see the video, but I can capture only first frame with `flv.js`.
Does `flv.js` expects `sps` and `pps` headers for every frame and cannot handle global header?

**Comment by jkuri** [comment]
@xqq after pointing me out I was setting `frame->pts` totally wrong I was playing around that value.
I now have almost working example, but still some lag in video. 
I am currently setting `frame->pts` as
```c++
frame->pts = (1.0 / fps) * 1000 * frame_pts++;
```
I am searching for right formula to calculate the frame PTS, can you please point me to the right direction?

I must say I would be struggling with the issue for a long time if you wouldn't me point me that there's an issue with `frame->pts`. Thanks alot about that, I really appreciate it.

Btw, I updated my repository with the latest code I have if you maybe have time to check it out.

**Comment by jkuri** [comment]
```c++
frame->pts += av_rescale_q(1, out_codec_ctx->time_base, out_stream->time_base);
```
seems to provide fluent video 👍 
I have about 1s delay in video but I am closing this issue as this problem is out of scope for that issue context. thanks!

**Comment by xqq** [comment]
Not only `PTS` but also `DTS` is needed to be calculated, in my opinion.

I think there's a more complicated algorithm to marking `DTS`(**DecodingTimeStamp**) and `PTS`(**PresentationTimeStamp**) for a video frame, especially if `B-frames` are existed.

I'm not very familiar with FFmpeg encoding API. It's better to reference other stream publish projects, like **Open Broadcast Software**.

**Comment by jkuri** [comment]
@xqq thank you for your help. After setting `enableStashBuffer` to `false` I got a smooth, non-latency video stream. [Here](https://stackoverflow.com/questions/48578088/streaming-flv-to-rtmp-with-ffmpeg-using-h264-codec-and-c-api-to-flv-js/48619330#48619330) is the link to stackoverflow question where I answered my own question and provided full code.
Interestingly I didn't noticed `Open Broadcast Software` before after all my searches but I'll definitely take a look at this project.
